source/COMRegistrationFixes/__init__.py

```diff
@@ -0,0 +1,68 @@
+import os
+import subprocess
+import winVersion
+from logHandler import log
+
+# Particular  64 bit / 32 bit system paths
+systemRoot=os.path.expandvars('%SYSTEMROOT%')
+system32=os.path.join(systemRoot,'system32')
+sysWow64=os.path.join(systemRoot,'syswow64')
+systemDrive=os.path.expandvars('%SYSTEMDRIVE%\\')
+programFiles=os.path.join(systemDrive,'program files')
+programFilesX86=os.path.join(systemDrive,'program files (x86)')
+
+def registerServer(fileName,wow64=False):
+	"""
+	Registers the COM proxy dll with the given file name
+	Using regsvr32.
+	@ param fileName: the path to the dll
+	@type fileName: str
+	@param wow64: If true then the 32 bit (wow64) version of regsvr32 will be used.
+	@type wow64: bool
+	"""
+	regsvr32=os.path.join(sysWow64 if wow64 else system32,'regsvr32.exe')
+	try:
+		subprocess.check_call([regsvr32,'/s',fileName])
+	except subprocess.CalledProcessError as e:
+		log.error("Error registering %s, %s"%(fileName,e))
+	else:
+		log.debug("Registered %s"%fileName)
+
+def applyRegistryPatch(fileName):
+	"""
+	Applies the registry patch with the given file name
+	using regedit.
+	@ param fileName: the path to the dll
+	@type fileName: str
+	"""
+	regedit=os.path.join(system32,'regedit.exe')
+	try:
+		subprocess.check_call([regedit,'/s',fileName])
+	except subprocess.CalledProcessError as e:
+		log.error("Error applying registry patch: %s, %s"%(fileName,e))
+	else:
+		log.debug("Applied registry patch: %s"%fileName)
+
+def fixCOMRegistrations():
+	"""
+	Registers most common COM proxies, in case they had accidentally been unregistered or overwritten by 3rd party software installs/uninstalls.
+	"""
+	is64bit=os.environ.get("PROCESSOR_ARCHITEW6432","").endswith('64')
+	OSMajorMinor=winVersion.winVersion[:2]
+	log.debug("Fixing COM registration for Windows %s.%s, %s"%(OSMajorMinor[0],OSMajorMinor[1],"64 bit" if is64bit else "32 bit"))  
+	# Commands taken from NVDA issue #2807 comment https://github.com/nvaccess/nvda/issues/2807#issuecomment-320149243
+	# OLEACC (MSAA) proxies
+	applyRegistryPatch(os.path.join('COMRegistryFixes','oleaccProxy.reg'))
+	if is64bit:
+		applyRegistryPatch(os.path.join('COMRegistryFixes','oleaccProxy64.reg'))
+	# IDispatch and other common OLE interfaces
+	registerServer(os.path.join(system32,'oleaut32.dll'))
+	registerServer(os.path.join(system32,'actxprxy.dll'))
+	if is64bit:
+		registerServer(os.path.join(sysWow64,'oleaut32.dll'),wow64=True)
+		registerServer(os.path.join(sysWow64,'actxprxy.dll'),wow64=True)
+	# IServiceProvider on windows 7 can become unregistered 
+	if OSMajorMinor==(6,1): # Windows 7
+		registerServer(os.path.join(programFiles,'Internet Explorer','ieproxy.dll'))
+		if is64bit:
+			registerServer(os.path.join(programFilesX86,'Internet Explorer','ieproxy.dll'),wow64=True)

```

Comment:
If its 64 bit windows 7 then register the ieproxy.dll from both "program files/internet explorer/" and "program files (x86)/internet explorer"?

Reply:
Correct. There are COM proxies for both 64 bit and 32 bit on 64 bit machines.  The first registration will be 32 bit on 32 bit machines and 64 bit on 64 bit machines. Then if the machine is 64 bit, we also need to register the 32 bit one, so that 32 bit apps (including NVDA) can work. Note that IServiceProvider is used in more places than just Internet Explorer, its just that MS chose to store it in ieproxy.dll on Windows 7.